### DISCUSSION TITLE
File preloading in parallel and in the background

### DISCUSSION THREAD

**Discussion by zhangchi199133** [jspsych/jsPsych#961] [Q&A]
Hi there!
I have an experiment with a large number of stimuli (400 images and 300 audio files). If I used automatic preload, it is possible that the experiment will fail to load because of the timeout. So I used manual reploading to load the files in my trial_stimuli list, following https://www.jspsych.org/core_library/jspsych-pluginAPI/#jspsychpluginapigetaudiobuffer. The codes are as follow:

```javascript
var preloadProcedure = {
  timeline_variables: trial_stimuli,
  timeline: [
    {
      type: "call-function",
      async: true,
      func: function (done) {
        const makePreloadAudioPromise = function (audios) {
          return new Promise(function (resolve) {
            jsPsych.pluginAPI.preloadAudioFiles(audios, resolve);
          });
        };
        const audioPromise = makePreloadAudioPromise([
          jsPsych.timelineVariable("audio", true),
        ]);
        audioPromise.then(function () {
          done({ status: "success" });
        });
      },
    },
  ],
};

```

This could work, but the disadvantage of it is clearly that such a non-parallel way of file loading takes a great amount of time (in some cases even more than fifteen minutes), which was not ideal at all. 

So I was wondering whether it is possible to improve the efficiency of preloading. The solution I can think of right now are 1) to maunally load multiple files in parallel and 2) to run preloading in the background. But I have no idea how either of them can be done. Would you please help me with this? Thank you so much!

Best regards,

Chi

**Comment by vijaymarupudi** [comment]
So you want something like this I think:

This is pseudo code

```javascript
startPreloadingBlocks()
  if (block1finished) {
    showBlock1()
  }
  if (block2finished) {
    showBlock2()
  }
  if (block3finished) {
    showBlock3()
  }
  // ... etc
```

We'll need the preloading function

```javascript
function makePreloadAudioPromise(audios) {
  return new Promise(function (resolve) {
    jsPsych.pluginAPI.preloadAudioFiles(audios, resolve);
  });
}
```

What you'll first need to do is split your stimuli into even sized blocks. I'm going to assume that you've done so using this structure (example has 2 blocks)


```javascript
const stimuli = [
 ['audio_file1.wav', 'audio_file2.wav'],
 ['audio_file3.wav', 'audio_file4.wav'],
]
```

You then need a way to check if a block has completed loading

```javascript
const loadingIndicators = [null, null]
```

We will use Javascript promises instead of `null` once we start preloading your stimuli.

Lets now make a function that preloads the blocks. It should preload within block concurrently, and between blocks serially.

```javascript
function preloadBlocks() {
  let previousIndicator = Promise.resolve();
  for (let i = 0; i < stimuli.length; i++) {
    loadingIndicators[i] = previousIndicator.then(function () {
      return makePreloadAudioPromise(stimuli[i]);
    });
    previousIndicator = loadingIndicators[i];
  }
}
```


At the beginning of the experiment, start preloading

```javascript
timeline.push({
  type: 'call-function',
  func: function() {
    preloadBlocks()
  }
})
```

Now before showing jsPsych trials for a block, use an async call-function plugin to ensure that the stimuli for the block have been loaded. Ideally the participant will only have to wait once, at the beginning of the experiment.


```javascript
timeline.push({
  type: "call-function",
  async: true,
  func: function (done) {
    jsPsych.getDisplayElement().innerHTML = "<p> The experiment is loading...</p>";
    // this is the first block, block 0. (Javascript uses 0 indexing)
    loadingIndicators[0].then(function () {
      done({ status: "success" });
    });
  },
});
```

Then show the stimuli for block 1 and repeat.

This uses Promises extensively, read this to understand how they work: https://javascript.info/promise-basics

**Reply by zhangchi199133** [reply]
Hi @vijaymarupudi Thank you so much for the very detailed explanation.

I tried to align to your codes, but it seemed not working. Maybe it was because that instead of an array I tried to preload files from an object `stimuli_loader `. 

Here are the code I used:

`    var my_stimuli = [
        trial_stimuli1,
        trial_stimuli2, trial_stimuli3, trial_stimuli4, trial_stimuli5, trial_stimuli6, trial_stimuli7, trial_stimuli8,
        trial_stimuli9, trial_stimuli10, trial_stimuli11, trial_stimuli12, trial_stimuli13, trial_stimuli14, trial_stimuli15, trial_stimuli16,
        trial_stimuli17, trial_stimuli18, trial_stimuli19, trial_stimuli20, trial_stimuli21, trial_stimuli22, trial_stimuli23, trial_stimuli24
                              ];
        var trial_stimuli = jsPsych.randomization.sampleWithReplacement(my_stimuli, 1)[0];

        const trial_stimuli_bin1 = trial_stimuli.slice(0, 59)
        const trial_stimuli_bin2 = trial_stimuli.slice(60, 119)
        const trial_stimuli_bin3 = trial_stimuli.slice(120, 179)

        let stimuli_loader = []
        stimuli_loader.push( trial_stimuli_bin1, trial_stimuli_bin2, trial_stimuli_bin3)

        function makePreloadAudioPromise(audios) {
          return new Promise(function (resolve) {
            jsPsych.pluginAPI.preloadAudioFiles(audios, resolve);
          });
        }
        const loadingIndicators = [null, null]

        function makePreloadAudioPromise(audios) {
          return new Promise(function (resolve) {
            jsPsych.pluginAPI.preloadAudioFiles(audios, resolve);
          });
        }

        function preloadBlocks() {
          let previousIndicator = Promise.resolve();
          for (let i = 0; i < Object.keys(stimuli_loader).length; i++) {
            loadingIndicators[i] = previousIndicator.then(function () {
              return makePreloadAudioPromise(stimuli_loader[i].audio);
            });
            previousIndicator = loadingIndicators[i];
          }
        }

        timeline.push({
          type: 'call-function',
          func: function() {
            preloadBlocks()
          }
        })

        timeline.push({
        type: "call-function",
        async: true,
        func: function (done) {
          jsPsych.getDisplayElement().innerHTML = "<p> The experiment is loading...</p>";
          // this is the first block, block 0. (Javascript uses 0 indexing)
          loadingIndicators[0].then(function () {
            done({ status: "success" });
          });
        },
      });`

Here is the error I got:
`jspsych.js:2611 Uncaught (in promise) TypeError: Cannot read property 'length' of undefined
    at Object.jsPsych.utils.module.flatten (jspsych.js:2611)
    at Object.jsPsych.pluginAPI.module.preloadAudioFiles (jspsych.js:2287)
    at experiment.js:617
    at new Promise (<anonymous>)
    at makePreloadAudioPromise (experiment.js:616)
    at experiment.js:625`

It pointed to the lines in my experiment.js:
line 616-618
`          return new Promise(function (resolve) {
            jsPsych.pluginAPI.preloadAudioFiles(audios, resolve);
          });`

and line 625
` return makePreloadAudioPromise(stimuli_loader[i].audio);`

I was wondering where might go wrong. Thanks a lot!

Best,

Chi

**Reply by vijaymarupudi** [reply]
Can you insert a `console.log` on the argument to the makePreloadAudioPromise function? It's possible that it is not what you think it is. `stimuli_loader[i]` will be an array, not an object with an audio property as you're using it currently.

**Reply by zhangchi199133** [reply]
Hi! Sorry for the late response. I used `console.log(typeof stimuli_loader)` command to check the type of stimuli_loader. It actually returned 'object'. The variable seems to me an array of objects, which I did not know how to classify.

**Comment by vijaymarupudi** [comment]
Right, in JavaScript, an Array in also an Object, the typeof operator is not useful.

The function was designed to takes an array of strings. You can use stimuli_loader[i].map(function(x){return x.audio;}) to convert to that, assuming the objects in the Array have an audio property with the filename.
